src/scalar.h

```diff
@@ -92,8 +92,9 @@ static int secp256k1_scalar_eq(const secp256k1_scalar *a, const secp256k1_scalar
 
 /** Find r1 and r2 such that r1+r2*2^128 = k. */
 static void secp256k1_scalar_split_128(secp256k1_scalar *r1, secp256k1_scalar *r2, const secp256k1_scalar *k);
-/** Find r1 and r2 such that r1+r2*lambda = k,
- * where r1 and r2 or their negations are maximum 128 bits long (see secp256k1_ge_mul_lambda). */
+/** Find r1 and r2 such that r1+r2*lambda = k, where r1 and r2 or their
+ *  negations are maximum 128 bits long (see secp256k1_ge_mul_lambda). It is
+ *  required that r1, r2, and k all point to different objects. */
```

Comment:
Fwiw, I had discussed this with @jonasnick somewhere else. 

TL;DR: The promise you give by `restrict` is more involved than we assumed. 

He had tried just `restrict k` but that it turns out that doesn't change anything at all (and the compiler won't complain about incorrect usage), because `*k` is not modified in the function. After looking up the standard and other explanation, at least I was pretty convinced that `restrict r1, restrict r2, restrict k` is correct and meaningful, but @jonasnick preferred not to use it to keep things simple.

The best simple but precise explanation we found is this:
> The restrict modifier means that the compiler can assume that the object to which that pointer points either is not modified or is not referenced in any other way in the function. If the object is modified, then all access to that object must be done through that pointer.
 (https://www.reddit.com/r/C_Programming/comments/1xqf90/comment/cfdoon2/?context=3)

I wasn't aware about the "modified" part at all, I had assumed `restrict` has only to do with references.

Reply:
Fwiw, I'm happy to ACK either solution.

Reply:
As far as I can see, if restrict is added and there's a violation (some pointers point to the same object), then
```
VERIFY_CHECK(r1 != k);
VERIFY_CHECK(r2 != k);
VERIFY_CHECK(r1 != r2);
```
may not actually abort the program because we told the compiler that these conditions will never happen. It's not guaranteed that the compiler will be able to determine that violation statically and warn us about it. On the other hand, if we don't add restrict then the VERIFY_CHECKs will always (in the tests) find violations of the function's contract.

Reply:
Ok, I added the restricts.